File: docs/verifier/verify.js
#!/usr/bin/env node
/**
 * DeathRoll Luck — public Node.js verifier.
 *
 * Zero-dependency, single-file implementation of the bot's
 * provably-fair algorithm. Same algorithm as ``verify.py``;
 * CI cross-checks them byte-for-byte against each other AND
 * against the bot's ``deathroll_core/fairness/`` package.
 *
 * Usage:
 *
 *   node verify.js <game> <server_seed_hex> <client_seed> <nonce> [extra...]
 *
 * Examples:
 *
 *   node verify.js coinflip d4...3fb cs 0
 *   node verify.js mines    cafe...    cs 0 3 25
 *   node verify.js blackjack cafe...   cs 0 6
 *
 * Output is JSON on stdout.
 */

const crypto = require('crypto');

// ---------------------------------------------------------------------------
// Engine
// ---------------------------------------------------------------------------

function compute(serverSeed, clientSeed, nonce) {
    const msg = Buffer.from(`${clientSeed}:${nonce}`, 'utf8');
    return crypto.createHmac('sha512', serverSeed).update(msg).digest();
}

function extend(out, byteCount) {
    if (byteCount <= out.length) return out.slice(0, byteCount);
    const chunks = [out];
    let total = out.length;
    let counter = 0;
    while (total < byteCount) {
        counter += 1;
        const counterBuf = Buffer.alloc(4);
        counterBuf.writeUInt32BE(counter, 0);
        const chunk = crypto
            .createHash('sha256')
            .update(Buffer.concat([out, counterBuf]))
            .digest();
        chunks.push(chunk);
        total += chunk.length;
    }
    return Buffer.concat(chunks).slice(0, byteCount);
}

// ---------------------------------------------------------------------------
// Byte-stream + Fisher-Yates helpers
// ---------------------------------------------------------------------------

function* byteStream(out) {
    for (const b of out) yield b;
    let counter = 0;
    while (true) {
        counter += 1;
        const counterBuf = Buffer.alloc(4);
        counterBuf.writeUInt32BE(counter, 0);
        const chunk = crypto
            .createHash('sha256')
            .update(Buffer.concat([out, counterBuf]))
            .digest();
        for (const b of chunk) yield b;
    }
}

function pullUint(stream, width) {
    let value = 0;
    for (let i = 0; i < width; i++) {
        value = value * 256 + stream.next().value;
    }
    return value;
}

function fisherYatesPartial(stream, n, k, pickWidth = 4) {
    const arr = [];
    for (let i = 0; i < n; i++) arr.push(i);
    const bound = Math.max(1, n - k);
    for (let i = n - 1; i >= bound; i--) {
        const j = pullUint(stream, pickWidth) % (i + 1);
        const tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }
    // settled slots are arr[n-k:n]; reverse so the first selected
    // is at index 0.
    return arr.slice(n - k).reverse();
}

// ---------------------------------------------------------------------------
// Decoders
// ---------------------------------------------------------------------------

function decodeCoinflip(out) {
    return (out[0] & 1) === 0 ? 'heads' : 'tails';
}

function decodeDice(out) {
    // BE int from first 4 bytes.
    const n =
        out[0] * 0x1000000 + out[1] * 0x10000 + out[2] * 0x100 + out[3];
    return (n % 10000) / 100;
}

function decode99x(out) {
    return (out[0] % 100) + 1;
}

function decodeHotcold(out) {
    const n = ((out[0] << 8) | out[1]) % 10000;
    if (n < 500) return 'rainbow';
    if (n < 5250) return 'hot';
    return 'cold';
}

function decodeRouletteEu(out) {
    return ((out[0] << 8) | out[1]) % 37;
}

function decodeMinesPositions(out, minesCount, gridSize) {
    if (!(1 <= minesCount && minesCount < gridSize)) {
        throw new Error(
            `mines_count out of range: ${minesCount} (must be in [1, ${gridSize - 1}])`,
        );
    }
    const stream = byteStream(out);
    return fisherYatesPartial(stream, gridSize, minesCount);
}

function decodeBlackjackDeck(out, decks) {
    if (decks <= 0) {
        throw new Error(`decks must be positive: ${decks}`);
    }
    const cards = [];
    for (let c = 0; c < 52; c++) {
        for (let d = 0; d < decks; d++) cards.push(c);
    }
    const stream = byteStream(out);
    const n = cards.length;
    for (let i = n - 1; i > 0; i--) {
        const j = pullUint(stream, 4) % (i + 1);
        const tmp = cards[i];
        cards[i] = cards[j];
        cards[j] = tmp;
    }
    return cards;
}

function decodeDiceDuel(out) {
    return [(out[0] % 12) + 1, (out[1] % 12) + 1];
}

function decodeStakingDuel(out, maxRounds) {
    if (maxRounds <= 0) {
        throw new Error(`max_rounds must be positive: ${maxRounds}`);
    }
    const stream = byteStream(out);
    const rounds = [];
    for (let r = 0; r < maxRounds; r++) {
        const p = (stream.next().value % 12) + 1;
        const b = (stream.next().value % 12) + 1;
        rounds.push({ player_roll: p, bot_roll: b });
    }
    return rounds;
}

function decodeRaffleWinners(out, ticketCount) {
    if (ticketCount < 3) {
        throw new Error(
            `need at least 3 tickets for 3 winners; got ${ticketCount}`,
        );
    }
    const stream = byteStream(out);
    return fisherYatesPartial(stream, ticketCount, 3);
}

// ---------------------------------------------------------------------------
// Dispatch
// ---------------------------------------------------------------------------

function dispatch(game, out, args) {
    switch (game) {
        case 'coinflip':
            return decodeCoinflip(out);
        case 'dice':
            return decodeDice(out);
        case '99x':
            return decode99x(out);
        case 'hotcold':
            return decodeHotcold(out);
        case 'roulette':
            return decodeRouletteEu(out);
        case 'diceduel':
            return decodeDiceDuel(out);
        case 'mines':
            if (args.length !== 2) throw new Error('mines: <mines_count> <grid_size>');
            return decodeMinesPositions(out, parseInt(args[0], 10), parseInt(args[1], 10));
        case 'blackjack':
            if (args.length !== 1) throw new Error('blackjack: <decks>');
            return decodeBlackjackDeck(out, parseInt(args[0], 10));
        case 'staking':
            if (args.length !== 1) throw new Error('staking: <max_rounds>');
            return decodeStakingDuel(out, parseInt(args[0], 10));
        case 'raffle':
            if (args.length !== 1) throw new Error('raffle: <ticket_count>');
            return decodeRaffleWinners(out, parseInt(args[0], 10));
        default:
            throw new Error(`unknown game: ${game}`);
    }
}

function main() {
    const argv = process.argv.slice(2);
    if (argv.length < 4) {
        console.error(
            'Usage: verify.js <game> <server_seed_hex> <client_seed> <nonce> [extra_args...]',
        );
        process.exit(2);
    }
    const game = argv[0];
    const serverSeed = Buffer.from(argv[1], 'hex');
    const clientSeed = argv[2];
    const nonce = parseInt(argv[3], 10);
    const extra = argv.slice(4);

    const head = compute(serverSeed, clientSeed, nonce);
    const extended = extend(head, 4096);
    const outcome = dispatch(game, extended, extra);
    process.stdout.write(JSON.stringify(outcome) + '\n');
}

main();

File: docs/verifier/verify.py
#!/usr/bin/env python3
"""DeathRoll Luck — public Python verifier.

Zero-dependency, single-file implementation of the bot's
provably-fair algorithm. Any user can clone this script and
re-derive the outcome of any past bet locally to confirm the
casino did not tamper.

The algorithm:

    out = HMAC-SHA512(
        key=server_seed,
        message=f"{client_seed}:{nonce}".encode("utf-8"),
    )

For games that need more than 64 bytes (Blackjack's full deck
shuffle of decks*52 cards), the byte stream is extended
deterministically via:

    chunk_n = SHA-256(out || n.to_bytes(4, 'big'))   for n = 1, 2, ...

This file is the canonical reference. The bot's
``deathroll_core/fairness/`` package implements the same
algorithm and CI cross-checks them byte-for-byte against this
script and against ``verify.js``.

Usage:

    python verify.py <game> <server_seed_hex> <client_seed> <nonce> [extra...]

Examples:

    python verify.py coinflip d4...3fb cs 0
    python verify.py mines    cafe...    cs 0 3 25
    python verify.py blackjack cafe...   cs 0 6
    python verify.py staking  cafe...    cs 0 5
    python verify.py raffle   cafe...    cs 0 100

Output is JSON on stdout.
"""

from __future__ import annotations

import hashlib
import hmac
import json
import sys
from typing import Any

# ---------------------------------------------------------------------------
# Engine
# ---------------------------------------------------------------------------


def compute(server_seed: bytes, client_seed: str, nonce: int) -> bytes:
    """HMAC-SHA512(server_seed, f'{client_seed}:{nonce}'). 64 bytes."""
    msg = f"{client_seed}:{nonce}".encode()
    return hmac.new(server_seed, msg, hashlib.sha512).digest()


def extend(out: bytes, byte_count: int) -> bytes:
    """Extend out via SHA-256 chain to byte_count bytes."""
    if byte_count <= len(out):
        return out[:byte_count]
    extended = bytearray(out)
    counter = 0
    while len(extended) < byte_count:
        counter += 1
        chunk = hashlib.sha256(out + counter.to_bytes(4, "big")).digest()
        extended.extend(chunk)
    return bytes(extended[:byte_count])


# ---------------------------------------------------------------------------
# Byte-stream + Fisher-Yates helpers (mirror deathroll_core/fairness/decoders)
# ---------------------------------------------------------------------------


def _byte_stream(out: bytes):
    """Lazy iterator: out + SHA-256(out || counter) chain."""
    yield from out
    counter = 0
    while True:
        counter += 1
        chunk = hashlib.sha256(out + counter.to_bytes(4, "big")).digest()
        yield from chunk


def _pull_uint(stream, width: int) -> int:
    value = 0
    for _ in range(width):
        value = (value << 8) | next(stream)
    return value


def _fisher_yates_partial(
    stream, n: int, k: int, *, pick_width: int = 4
) -> list[int]:
    """First-k of FY shuffle of [0, n)."""
    arr = list(range(n))
    bound = max(1, n - k)
    for i in range(n - 1, bound - 1, -1):
        j = _pull_uint(stream, pick_width) % (i + 1)
        arr[i], arr[j] = arr[j], arr[i]
    return arr[n - k :][::-1]


# ---------------------------------------------------------------------------
# Decoders
# ---------------------------------------------------------------------------


def decode_coinflip(out: bytes) -> str:
    return "heads" if (out[0] & 1) == 0 else "tails"


def decode_dice(out: bytes) -> float:
    n = int.from_bytes(out[:4], "big")
    return (n % 10000) / 100


def decode_99x(out: bytes) -> int:
    return (out[0] % 100) + 1


def decode_hotcold(out: bytes) -> str:
    n = int.from_bytes(out[:2], "big") % 10000
    if n < 500:
        return "rainbow"
    if n < 5250:
        return "hot"
    return "cold"


def decode_roulette_eu(out: bytes) -> int:
    return int.from_bytes(out[:2], "big") % 37


def decode_mines_positions(
    out: bytes, mines_count: int, grid_size: int
) -> list[int]:
    if not (1 <= mines_count < grid_size):
        raise ValueError(
            f"mines_count out of range: {mines_count} "
            f"(must be in [1, {grid_size - 1}])"
        )
    stream = _byte_stream(out)
    return _fisher_yates_partial(stream, n=grid_size, k=mines_count)


def decode_blackjack_deck(out: bytes, decks: int) -> list[int]:
    if decks <= 0:
        raise ValueError(f"decks must be positive: {decks}")
    cards = [c for c in range(52) for _ in range(decks)]
    stream = _byte_stream(out)
    n = len(cards)
    for i in range(n - 1, 0, -1):
        j = _pull_uint(stream, 4) % (i + 1)
        cards[i], cards[j] = cards[j], cards[i]
    return cards


def decode_dice_duel(out: bytes) -> tuple[int, int]:
    return ((out[0] % 12) + 1, (out[1] % 12) + 1)


def decode_staking_duel(out: bytes, max_rounds: int) -> list[dict[str, int]]:
    if max_rounds <= 0:
        raise ValueError(f"max_rounds must be positive: {max_rounds}")
    stream = _byte_stream(out)
    rounds: list[dict[str, int]] = []
    for _ in range(max_rounds):
        p = (next(stream) % 12) + 1
        b = (next(stream) % 12) + 1
        rounds.append({"player_roll": p, "bot_roll": b})
    return rounds


def decode_raffle_winners(out: bytes, ticket_count: int) -> list[int]:
    if ticket_count < 3:
        raise ValueError(
            f"need at least 3 tickets for 3 winners; got {ticket_count}"
        )
    stream = _byte_stream(out)
    return _fisher_yates_partial(stream, n=ticket_count, k=3)


# ---------------------------------------------------------------------------
# Dispatch
# ---------------------------------------------------------------------------


def _decode(game: str, out: bytes, args: list[str]) -> Any:
    if game == "coinflip":
        return decode_coinflip(out)
    if game == "dice":
        return decode_dice(out)
    if game == "99x":
        return decode_99x(out)
    if game == "hotcold":
        return decode_hotcold(out)
    if game == "roulette":
        return decode_roulette_eu(out)
    if game == "diceduel":
        return list(decode_dice_duel(out))
    if game == "mines":
        if len(args) != 2:
            raise SystemExit("mines: <mines_count> <grid_size>")
        return decode_mines_positions(out, int(args[0]), int(args[1]))
    if game == "blackjack":
        if len(args) != 1:
            raise SystemExit("blackjack: <decks>")
        return decode_blackjack_deck(out, int(args[0]))
    if game == "staking":
        if len(args) != 1:
            raise SystemExit("staking: <max_rounds>")
        return decode_staking_duel(out, int(args[0]))
    if game == "raffle":
        if len(args) != 1:
            raise SystemExit("raffle: <ticket_count>")
        return decode_raffle_winners(out, int(args[0]))
    raise SystemExit(f"unknown game: {game}")


def main() -> int:
    if len(sys.argv) < 5:
        print(
            "Usage: verify.py <game> <server_seed_hex> <client_seed> <nonce> "
            "[extra_args...]",
            file=sys.stderr,
        )
        return 2
    game = sys.argv[1]
    server_seed = bytes.fromhex(sys.argv[2])
    client_seed = sys.argv[3]
    nonce = int(sys.argv[4])
    extra = sys.argv[5:]

    head = compute(server_seed, client_seed, nonce)
    # Extend liberally — most decoders don't need it, blackjack needs ~1244.
    extended = extend(head, 4096)
    outcome = _decode(game, extended, extra)
    json.dump(outcome, sys.stdout)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())
